test/integration/apiserver/apiserver_test.go

```diff
@@ -2216,19 +2216,26 @@ func TestDedupOwnerReferences(t *testing.T) {
 			}
 			klog.Infof("creating dependent with duplicate owner references")
 			dependent := c.createDependentWithOwners([]metav1.OwnerReference{fakeRefA, fakeRefA})
+			assertManagedFields(t, dependent)
 			expectedWarning := fmt.Sprintf(handlers.DuplicateOwnerReferencesWarningFormat, fakeRefA.UID)
 			assertOwnerReferences(t, dependent, []metav1.OwnerReference{fakeRefA})
 			assertWarningCount(t, warningWriter, previousWarningCount+1)
 			assertWarningMessage(t, b, expectedWarning)
 
 			klog.Infof("updating dependent with duplicate owner references")
 			dependent = c.updateDependentWithOwners(dependent, []metav1.OwnerReference{fakeRefA, fakeRefA})
+			assertManagedFields(t, dependent)
 			assertOwnerReferences(t, dependent, []metav1.OwnerReference{fakeRefA})
 			assertWarningCount(t, warningWriter, previousWarningCount+2)
 			assertWarningMessage(t, b, expectedWarning)
 
 			klog.Infof("patching dependent with duplicate owner reference")
 			dependent = c.patchDependentWithOwner(dependent, fakeRefA)
+			// TODO: currently a patch request that duplicates owner references can still
+			// wipe out managed fields. Note that this happens to built-in resources but
+			// not custom resources. In future we should either dedup before writing manage
+			// fields, or stop deduping and reject the request.
+			// assertManagedFields(t, dependent)
```

Comment:
Two behaviors regarding managedFields that surprised me, but may be expected. FYI @apelisse 
1. once an object has empty managedFields, the field manager won't write managedFields for future requests made by this client, even if those requests don't contain duplicate owner refs. 
2. without this PR, the field manager still succeeded to write `managedFields` for custom resources, even when duplicate owner ref existed in the request. Note that `"f:ownerReferences"` in the custom resource doesn't contain UIDs. 

with this PR, for built-in resource:
```
# create with dup owner refs
managedFields: [{apiserver.test Update v1 2020-11-18 19:44:36 -0800 PST FieldsV1 {"f:metadata":{"f:ownerReferences":{".":{},"k:{\"uid\":\"03d3ba47-a815-46aa-84c7-70ee450b7761\"}":{".":{},"f:apiVersion":{},"f:kind":{},"f:name":{},"f:uid":{}}}}}}]
# update with dup owner refs
managedFields: [{apiserver.test Update v1 2020-11-18 19:44:36 -0800 PST FieldsV1 {"f:metadata":{"f:ownerReferences":{".":{},"k:{\"uid\":\"03d3ba47-a815-46aa-84c7-70ee450b7761\"}":{".":{},"f:apiVersion":{},"f:kind":{},"f:name":{},"f:uid":{}}}}}}]
# patch, append a dup owner ref
fieldmanager.go:186] [SHOULD NOT HAPPEN] failed to update managedFields for /, Kind=: failed to convert new object (/v1, Kind=ConfigMap) to smd typed: .metadata.ownerReferences: duplicate entries for key [uid="03d3ba47-a815-46aa-84c7-70ee450b7761"]
managedFields: []
# update, append a different owner ref
managedFields: []
# patch, append a different owner ref
managedFields: []
```
for custom resource (note the UID isn't written in the managed fields): 
```
# repeat all the steps above, the managed fields always read:
[{apiserver.test Update cr.bar.com/v1 2020-11-18 19:44:36 -0800 PST FieldsV1 {"f:metadata":{"f:ownerReferences":{}}}}]
```
without this PR, the behavior of built-in resources will be:
```
# create with dup owner refs
fieldmanager.go:186] [SHOULD NOT HAPPEN] failed to update managedFields for /, Kind=: failed to convert new object (/v1, Kind=ConfigMap) to smd typed: .metadata.ownerReferences: duplicate entries for key [uid="29b54e45-e089-4b12-bfd6-6ab63f20204a"]
managedFields: []
# repeat all the remaining steps:
managedFields: []
```
and the behavior of custom resources won't change.

Reply:
I didn't expect the managedFields to stay empty. Does this also happen on apply?

Reply:
> 1. once an object has empty managedFields, the field manager won't write managedFields for future requests made by this client, even if those requests don't contain duplicate owner refs. 

That's because we don't track fields for objects that don't have managed fields, until they are server-side applied. That's to prevent existing objects from suddenly becoming bigger (because of the managed-fields) on existing clusters.
When we wipe, we could decide that we need to keep tracking, but I'd say this is WAI.

> 2. without this PR, the field manager still succeeded to write `managedFields` for custom resources, even when duplicate owner ref existed in the request. Note that `"f:ownerReferences"` in the custom resource doesn't contain UIDs. 

it looks like ownerReferences is considered atomic for custom-resources. Can you look at the CRD and say if the ownerReference list is associative?

Reply:
> but I'd say this is WAI.

Thanks for the explanation! That makes sense. 

> it looks like ownerReferences is considered atomic for custom-resources

IIUC the field manager uses the CRD schema to determine that. The [CRD used](https://github.com/kubernetes/kubernetes/blob/18099e1ef7283d9ab09c45c5a4a90e26fdce1161/test/integration/etcd/data.go#L593-L606) doesn't have schema. I also verified that API server doesn't insert the in-tree object meta schema into CRD. I think this is expected.

Reply:
> The CRD used doesn't have schema.

We treat all lists as atomic for schema-less CRDs, all good! 👍